README.md
# NAME

Shuvgey - AnyEvent HTTP/2 (RFC 7540) Server for PSGI

# SYNOPSIS

    shuvgey --listen :8000 --tls_key=cert.key --tls_crt=cert.crt app.psgi

# DESCRIPTION

Shuvgey is a lightweight non-blocking, single-threaded HTTP/2 (RFC 7540) Server
that runs PSGI applications on top of [AnyEvent](https://metacpan.org/pod/AnyEvent) event loop.

Shuvgey use [Protocol::HTTP2](https://metacpan.org/pod/Protocol::HTTP2) for HTTP/2 support. Supported plain text HTTP/2
connections, HTTP/1.1 Upgrade, and secure TLS connections (with ALPN/NPN
protocol negotiation).

# STATUS

It's alpha stage. I can run simple [Dancer](https://metacpan.org/pod/Dancer) PSGI app and it even work!

# NAMING

There is a wellknown python non-blocking, single-threaded HTTP server Tornado.

Shuvgey is the collective name of evil forces in Komi-Zyryan and Komi-Perm
folklore. Materialized in the form of a strong wind vortex. See also wikipedia
article
[Шувгей](http://ru.wikipedia.org/wiki/%D0%A8%D1%83%D0%B2%D0%B3%D0%B5%D0%B9) (in
russian).

So Shuvgey is like Tornado, but more scary: written in Perl and support HTTP/2
protocol.

# OPTIONS

Avaliable all options from [plackup](https://metacpan.org/pod/plackup) and also some specific Shuvgey options:

- --no\_tls - don't encrypt connection
- --upgrade - use HTTP/1.1 Upgrade protocol to upgrade to HTTP/2 (no tls)
- --tls\_key - path to private key
- --tls\_crt - path to certificate

# LICENSE

Copyright (C) Vladimir Lettiev.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

# AUTHOR

Vladimir Lettiev <thecrux@gmail.com>


Changes
Revision history for Perl extension Shuvgey

{{$NEXT}}
    - fixed ALPN negotiation
    - dropped h2-14 compatibility

0.09 2015-07-09T21:24:41Z
    - required Protocol::HTTP2 >= 1.03 (improved conformance with RFC 7540)
    - required Plack >= 1.0037 (now accept HTTP/2 value for
      SERVER_PROTOCOL)
    - Shuvgey now pass all tests of
      [h2spec tool](https://github.com/summerwind/h2spec) (version 1.01) for
      conformance with HTTP/2 specification

0.08 2015-06-08T18:01:27Z
    - required Protocol::HTTP2 >= 1.00
    - preserved compatibility with h2-14 clients
    - setup MAX_CONCURRENT_STREAMS to 100

0.07 2015-02-11T14:05:36Z
    - HTTP/2 draft 17

0.06 2014-12-01T13:09:32Z
    - HTTP/2 draft 16
    - announce draft 14 support for interoperability
    - only TLS v1.2 allowed for secure connections

0.05 2014-10-28T12:50:53Z
    - slightly changed debug timestamps
    - return 505 error if client doesn't support HTTP/2
    - HTTP/2 draft 15

0.04 2014-09-09T13:09:10Z
    - added support for ECDH ciphers (Net-SSLeay >= 1.56)
    - required Protocol::HTTP2 >= 0.11 (HTTP/2 draft 14, HPACK draft 09)

0.03 2014-07-08T13:22:45Z
    - use system hostname
    - required Protocol::HTTP2 >= 0.09 (HTTP/2 draft 13, HPACK draft 08)

0.02 2014-05-17T10:19:39Z

    - filling PSGI request env
    - handling PSGI response
    - 500 Internal Error page
    - no_tls and upgrade options

0.01 2014-05-17T10:15:21Z

    - original version


lib/Shuvgey.pm
package Shuvgey;
use 5.008005;
use strict;
use warnings;

our $VERSION = "0.08";

1;
__END__

=encoding utf-8

=head1 NAME

Shuvgey - AnyEvent HTTP/2 (RFC 7540) Server for PSGI

=head1 SYNOPSIS

    shuvgey --listen :8000 --tls_key=cert.key --tls_crt=cert.crt app.psgi

=head1 DESCRIPTION

Shuvgey is a lightweight non-blocking, single-threaded HTTP/2 (RFC 7540) Server
that runs PSGI applications on top of L<AnyEvent> event loop.

Shuvgey use L<Protocol::HTTP2> for HTTP/2 support. Supported plain text HTTP/2
connections, HTTP/1.1 Upgrade, and secure TLS connections (with ALPN/NPN
protocol negotiation).

=head1 STATUS

It's alpha stage. I can run simple L<Dancer> PSGI app and it even work!

Shuvgey pass all tests of L<h2spec tool|https://github.com/summerwind/h2spec>
(version 1.01) for conformance with HTTP/2 implementation. Check it yourself:

    $ h2spec -p 443 -h shuvgey.net -t

=head1 NAMING

There is a wellknown python non-blocking, single-threaded HTTP server Tornado.

Shuvgey is the collective name of evil forces in Komi-Zyryan and Komi-Perm
folklore. Materialized in the form of a strong wind vortex. See also wikipedia
article
L<Шувгей|http://ru.wikipedia.org/wiki/%D0%A8%D1%83%D0%B2%D0%B3%D0%B5%D0%B9> (in
russian).

So Shuvgey is like Tornado, but more scary: written in Perl and support HTTP/2
protocol.

=head1 OPTIONS

Avaliable all options from L<plackup> and also some specific Shuvgey options:

=over

=item --no_tls - don't encrypt connection

=item --upgrade - use HTTP/1.1 Upgrade protocol to upgrade to HTTP/2 (no tls)

=item --tls_key - path to private key

=item --tls_crt - path to certificate

=back

=head1 LICENSE

Copyright (C) Vladimir Lettiev.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=head1 AUTHOR

Vladimir Lettiev E<lt>thecrux@gmail.comE<gt>

=cut



lib/Shuvgey/Server.pm
package Shuvgey::Server;
use strict;
use warnings;
use Net::SSLeay;
use AnyEvent;
use AnyEvent::Socket;
use AnyEvent::Handle;
use AnyEvent::TLS;
use Protocol::HTTP2;
use Protocol::HTTP2::Constants qw(const_name :settings);
use Protocol::HTTP2::Server;
use Data::Dumper;
use URI::Escape qw(uri_unescape);
use Carp;
use Sys::Hostname;
use Scalar::Util qw(blessed);
require Shuvgey;

use constant {
    TRUE  => !undef,
    FALSE => !!undef,

    STOP => exists $ENV{SHUVGEY_DEBUG},

    # Log levels
    DEBUG     => 0,
    INFO      => 1,
    NOTICE    => 2,
    WARNING   => 3,
    ERROR     => 4,
    CRITICAL  => 5,
    ALERT     => 6,
    EMERGENCY => 7,
};

my $start_time = 0;
my $hostname   = hostname;

sub talk($$) {
    if ( shift() >= $ENV{SHUVGEY_DEBUG} ) {
        my $message = shift;
        chomp($message);
        my $now = AnyEvent->now;
        if ( $now - $start_time < 60 ) {
            $message =~ s/\n/\n           /g;
            printf "[%05.3f] %s\n", $now - $start_time, $message;
        }
        else {
            my @t = ( localtime() )[ 5, 4, 3, 2, 1, 0 ];
            $t[0] += 1900;
            $t[1]++;
            $message =~ s/\n/\n                      /g;
            printf "[%4d-%02d-%02d %02d:%02d:%02d] %s\n", @t, $message;
            $start_time = $now;
        }
    }
}

sub new {
    my $class = shift;
    bless {@_}, $class;
}

sub run {
    my ( $self, $app ) = @_;

    STOP and talk DEBUG, Dumper($self);

    my ( $host, $port );
    if ( $self->{listen} ) {
        ( $self->{host}, $self->{port} ) = split /:/,
          shift @{ $self->{listen} };
    }

    $host = $self->{host} || undef;
    $port = $self->{port} || undef;

    if ( !exists $self->{no_tls} ) {
        STOP and talk INFO, "tls\n";
        if ( exists $self->{upgrade} ) {
            $self->{no_tls} = TRUE;
        }
        elsif (!exists $self->{tls_crt}
            || !-r $self->{tls_crt}
            || !exists $self->{tls_key}
            || !-r $self->{tls_key} )
        {
            croak "Can't read tls_crt or tls_key files\n"
              . "Use --no_tls option to start Shuvgey without tls";
        }
    }

    $self->{exit} = AnyEvent->condvar;

    $self->run_tcp_server( $app, $host, $port );

    my $w;
    $w = AnyEvent->signal(
        signal => "TERM",
        cb     => sub {
            undef $w;
            $self->finish("Received SIGTERM");
        }
    );

    my $recv = $self->{exit}->recv;
    STOP and talk INFO, $recv;
}

sub run_tcp_server {
    my ( $self, $app, $host, $port ) = @_;

    tcp_server $host, $port, sub {

        my ( $fh, $peer_host, $peer_port ) = @_;
        my $tls;

        STOP and talk INFO, "Accept connection from $peer_host:$peer_port";

        if ( !exists $self->{no_tls} ) {
            $tls = $self->create_tls or return;
        }

        my $handle;
        $handle = AnyEvent::Handle->new(
            fh       => $fh,
            autocork => 1,
            $tls
            ? (
                tls         => "accept",
                tls_ctx     => $tls,
                on_starttls => sub {
                    my ( $handle, $success, $error_message ) = @_;
                    if ( !$success ) {
                        STOP and talk ERROR, $error_message;
                    }
                    else {
                        my $proto = (
                            exists &Net::SSLeay::P_alpn_selected
                            ? Net::SSLeay::P_alpn_selected( $handle->{tls} )
                            : Net::SSLeay::P_next_proto_negotiated(
                                $handle->{tls}
                            )
                          )
                          || '';
                        STOP
                          and talk INFO, "Client negotiated protocol: $proto";
                        if ( $proto ne Protocol::HTTP2::ident_tls ) {
                            STOP and talk ERROR, "$proto not supported";
                            $handle->push_write( $self->error_505 );
                            $handle->push_shutdown;
                        }
                        else {
                            STOP and talk INFO, "tls started ok";
                            STOP
                              and talk INFO, "cipher: "
                              . Net::SSLeay::get_cipher( $handle->{tls} );
                            $self->start_server( $handle, $app, $host, $port,
                                $peer_host, $peer_port );
                        }
                    }
                },
                on_stoptls => sub {
                    my ($handle) = @_;
                    STOP and talk INFO, "tls stoped: <$!>";
                },
              )
            : (),
            on_error => sub {
                $_[0]->destroy;
                STOP and talk ERROR, "connection error";
            },
            on_eof => sub {
                $handle->destroy;
            }
        );
        $self->start_server( $handle, $app, $host, $port, $peer_host,
            $peer_port )
          unless $tls;
      },

      # Bound to host:port
      sub {
        ( undef, $host, $port ) = @_;
        STOP and talk NOTICE, "Ready to serve request\n";

        # For Plack::Runner
        $self->{server_ready}->(
            {
                host            => $host,
                port            => $port,
                server_software => 'Shuvgey',
            }
        ) if $self->{server_ready};
        return TRUE;
      };

    return TRUE;
}

sub start_server {
    my ( $self, $handle, $app, $host, $port, $peer_host, $peer_port ) = @_;
    my $server;

    $server = Protocol::HTTP2::Server->new(
        exists $self->{upgrade} ? ( upgrade => 1 ) : (),
        settings        => { &SETTINGS_MAX_CONCURRENT_STREAMS => 100, },
        on_change_state => sub {
            my ( $stream_id, $previous_state, $current_state ) = @_;
        },
        on_error => sub {
            my $error = shift;
            STOP and talk
              ERROR,
              sprintf "Error occured: %s\n",
              const_name( "errors", $error );
        },
        on_request => sub {
            my ( $stream_id, $headers, $data ) = @_;

            my $env =
              $self->psgi_env( $host, $port, $peer_host, $peer_port,
                $headers, $data );

            my $response = eval { $app->($env) }
              || $self->internal_error($@);

            # TODO: support for CODE
            if ( ref $response ne 'ARRAY' ) {
                $response =
                  $self->internal_error("PSGI CODE response not supported yet");
            }

            my $body;

            if ( ref $response->[2] eq 'ARRAY' ) {
                $body = join '', @{ $response->[2] };
            }
            elsif (
                ref $response->[2] eq 'GLOB'
                or ( blessed( $response->[2] )
                    && $response->[2]->can('getline') )
              )
            {
                local $/ = \4096;
                $body = '';
                while ( defined( my $chunk = $response->[2]->getline ) ) {
                    $body .= $chunk;
                }
            }
            else {
                STOP and talk INFO, Dumper $response->[2];
                $response =
                  $self->internal_error( "body ref type "
                      . ( ref $response->[2] )
                      . " not supported yet" );
            }

            my @h = ();
            while ( @{ $response->[1] } ) {
                my ( $h, $v ) = splice @{ $response->[1] }, 0, 2;
                STOP and talk INFO, $h . " = " . $v;
                push @h, $h, $v unless lc($h) eq 'server';
            }
            push @h, "Server", "Shuvgey/" . $Shuvgey::VERSION;

            $server->response(
                stream_id => $stream_id,
                ':status' => $response->[0],
                headers   => \@h,
                data      => $body,
            );
        },
    );

    # First send settings to peer
    while ( my $frame = $server->next_frame ) {
        $handle->push_write($frame);
    }

    $handle->on_read(
        sub {
            my $handle = shift;

            $server->feed( $handle->{rbuf} );

            $handle->{rbuf} = undef;
            while ( my $frame = $server->next_frame ) {
                $handle->push_write($frame);
            }
            $handle->push_shutdown if $server->shutdown;
        }
    );
    ();
}

sub create_tls {
    my $self = shift;
    my $tls;

    eval {
        $tls = AnyEvent::TLS->new(
            method    => "TLSv1_2",
            cert_file => $self->{tls_crt},
            key_file  => $self->{tls_key},
        );

        # ECDH curve ( Net-SSLeay >= 1.56, openssl >= 1.0.0 )
        if ( exists &Net::SSLeay::CTX_set_tmp_ecdh ) {
            my $curve = Net::SSLeay::OBJ_txt2nid('prime256v1');
            my $ecdh  = Net::SSLeay::EC_KEY_new_by_curve_name($curve);
            Net::SSLeay::CTX_set_tmp_ecdh( $tls->ctx, $ecdh );
            Net::SSLeay::EC_KEY_free($ecdh);
        }

        # ALPN (Net-SSLeay > 1.55, openssl >= 1.0.2)
        if ( exists &Net::SSLeay::CTX_set_alpn_select_cb ) {
            Net::SSLeay::CTX_set_alpn_select_cb( $tls->ctx,
                [Protocol::HTTP2::ident_tls] );
        }

        # NPN  (Net-SSLeay > 1.45, openssl >= 1.0.1)
        elsif ( exists &Net::SSLeay::CTX_set_next_protos_advertised_cb ) {
            Net::SSLeay::CTX_set_next_protos_advertised_cb( $tls->ctx,
                [Protocol::HTTP2::ident_tls] );
        }
        else {
            die "ALPN and NPN are not supported\n";
        }
    };

    $self->finish("Some problem with TLS: $@\n") if $@;
    return $tls;
}

sub finish {
    shift->{exit}->send(shift);
}

sub psgi_env {
    my ( $self, $host, $port, $peer_host, $peer_port, $headers, $data ) = @_;

    my $input;
    open $input, '<', \$data if defined $data;

    my $env = {
        'psgi.version'      => [ 1, 1 ],
        'psgi.input'        => $input,
        'psgi.errors'       => *STDERR,
        'psgi.multithread'  => FALSE,
        'psgi.multiprocess' => FALSE,
        'psgi.run_once'     => FALSE,
        'psgi.nonblocking'  => TRUE,
        'psgi.streaming'    => FALSE,
        'SCRIPT_NAME'       => '',
        'SERVER_NAME' => $host eq '0.0.0.0' ? $hostname : $host,
        'SERVER_PORT' => $port,

        'SERVER_PROTOCOL' => "HTTP/2",

        # This not in PSGI spec. Why not?
        'REMOTE_HOST' => $peer_host,
        'REMOTE_ADDR' => $peer_host,
        'REMOTE_PORT' => $peer_port,
    };

    for my $i ( 0 .. @$headers / 2 - 1 ) {
        my ( $h, $v ) = ( $headers->[ $i * 2 ], $headers->[ $i * 2 + 1 ] );
        if ( $h eq ':method' ) {
            $env->{REQUEST_METHOD} = $v;
        }
        elsif ( $h eq ':scheme' ) {
            $env->{'psgi.url_scheme'} = $v;
        }
        elsif ( $h eq ':path' ) {
            $env->{REQUEST_URI} = $v;
            my ( $path, $query ) = ( $v =~ /^([^?]*)\??(.*)?$/s );
            $env->{QUERY_STRING} = $query || '';
            $env->{PATH_INFO} = uri_unescape($path);
        }
        elsif ( $h eq ':authority' ) {

            #TODO: what to do with :authority?
        }
        elsif ( $h eq 'content-length' ) {
            $env->{CONTENT_LENGTH} = $v;
        }
        elsif ( $h eq 'content-type' ) {
            $env->{CONTENT_TYPE} = $v;
        }
        else {
            my $header = 'HTTP_' . uc($h);
            if ( exists $env->{$header} ) {
                $env->{$header} .= ', ' . $v;
            }
            else {
                $env->{$header} = $v;
            }
        }
    }
    @$headers = ();
    STOP and talk INFO, Dumper($env);
    return $env;
}

sub internal_error {
    my ( $self, $error ) = @_;

    my $message = "500 - Internal Server Error";
    STOP and talk ERROR, "$message: $error\n";

    return [
        500,
        [
            'Content-Type'   => 'text/plain',
            'Content-Length' => length($message)
        ],
        [$message]
    ];
}

sub error_505 {
    my $error = "Shuvgey supports only HTTP/2 protocol";
    join "\x0d\x0a",
      "HTTP/1.1 505 HTTP version not supported",
      "Content-Type: text/plain",
      "Content-Length: " . length($error),
      "", $error;
}

1;
